pkg/services/sqlstore/dashboard_version.go

```diff
@@ -69,68 +70,47 @@ func GetDashboardVersions(query *m.GetDashboardVersionsQuery) error {
 
 func DeleteExpiredVersions(cmd *m.DeleteExpiredVersionsCommand) error {
 	return inTransaction(func(sess *DBSession) error {
-		versions := []DashboardVersionExp{}
-		versionsToKeep := setting.DashboardVersionsToKeep
+		const MAX_VERSIONS_TO_DELETE = 100
 
+		versionsToKeep := setting.DashboardVersionsToKeep
 		if versionsToKeep < 1 {
 			versionsToKeep = 1
 		}
 
-		err := sess.Table("dashboard_version").
-			Select("dashboard_version.id, dashboard_version.version, dashboard_version.dashboard_id").
-			Where(`dashboard_id IN (
-			SELECT dashboard_id FROM dashboard_version
-			GROUP BY dashboard_id HAVING COUNT(dashboard_version.id) > ?
-		)`, versionsToKeep).
-			Desc("dashboard_version.dashboard_id", "dashboard_version.version").
-			Find(&versions)
-
+		// Idea of this query is finding version IDs to delete based on formula:
+		// min_version_to_keep = min_version + (versions_count - versions_to_keep)
+		// where version stats is processed for each dashboard. This guarantees that we keep at least versions_to_keep
+		// versions, but in some cases (when versions are sparse) this number may be more.
+		versionIdsToDeleteSybqueryTemplate := `SELECT id
+			FROM dashboard_version, (
+				SELECT dashboard_id, count(version) as count, min(version) as min
+				FROM dashboard_version
+				GROUP BY dashboard_id
+				) AS vtd
+				WHERE dashboard_version.dashboard_id=vtd.dashboard_id
+				AND version < vtd.min + vtd.count - %v`
+
```

Comment:
Can we use a `LIMIT` clause here rather than loading everything only to discard it?

Reply:
I wanted to do that, but `LIMIT` is not a cross-platform syntax. MSSQL doesn't support it, for example. Using go-xorm `Limit()` is not possible, because it doesn't work with `SQL(expr)` function and I cannot split query into xorm functions.